source/dub/packagemanager.d

```diff
@@ -445,138 +494,126 @@ class PackageManager {
+	}
+	void removeLocalPackage(Path path, Path repoPath)
+	in { assert(repoPath.absolute); }
+	body
+	{
+		repoPath.normalize();
+		m_repos[repoPath].removeLocalPackage(path);
 	}
 
 	/// For the given type add another path where packages will be looked up.
-	void addSearchPath(Path path, LocalPackageType type)
+	deprecated void addSearchPath(Path path, LocalPackageType type)
 	{
-		m_repositories[type].searchPath ~= path;
-		writeLocalPackageList(type);
+		addSearchPath(path, m_defaultPaths[type]);
+	}
+	/// ditto
+	void addSearchPath(Path path, Path repoPath)
+	in { assert(repoPath.absolute); }
+	body
+	{
+		repoPath.normalize();
+		m_repos[repoPath].addSearchPath(path);
 	}
 
 	/// Removes a search path from the given type.
-	void removeSearchPath(Path path, LocalPackageType type)
+	deprecated void removeSearchPath(Path path, LocalPackageType type)
+	{
+		removeSearchPath(path, m_defaultPaths[type]);
+	}
+	/// ditto
+	void removeSearchPath(Path path, Path repoPath)
+	in { assert(repoPath.absolute); }
+	body
 	{
-		m_repositories[type].searchPath = m_repositories[type].searchPath.filter!(p => p != path)().array();
-		writeLocalPackageList(type);
+		repoPath.normalize();
+		m_repos[repoPath].removeSearchPath(path);
 	}
 
 	void refresh(bool refresh_existing_packages)
 	{
 		logDiagnostic("Refreshing local packages (refresh existing: %s)...", refresh_existing_packages);
 
 		// load locally defined packages
-		void scanLocalPackages(LocalPackageType type)
-		{
-			Path list_path = m_repositories[type].packagePath;
-			Package[] packs;
-			Path[] paths;
-			if (!m_disableDefaultSearchPaths) try {
-				auto local_package_file = list_path ~ LocalPackagesFilename;
-				logDiagnostic("Looking for local package map at %s", local_package_file.toNativeString());
-				if( !existsFile(local_package_file) ) return;
-				logDiagnostic("Try to load local package map at %s", local_package_file.toNativeString());
-				auto packlist = jsonFromFile(list_path ~ LocalPackagesFilename);
-				enforce(packlist.type == Json.Type.array, LocalPackagesFilename~" must contain an array.");
-				foreach( pentry; packlist ){
-					try {
-						auto name = pentry["name"].get!string;
-						auto path = Path(pentry["path"].get!string);
-						if (name == "*") {
-							paths ~= path;
-						} else {
-							auto ver = Version(pentry["version"].get!string);
-
-							Package pp;
-							if (!refresh_existing_packages) {
-								foreach (p; m_repositories[type].localPackages)
-									if (p.path == path) {
-										pp = p;
-										break;
-									}
-							}
+		if (!m_disableRepoSearchPaths)
+			over_repos: foreach (ref repo; m_repos)
+			{
+				Path list_path = repo.packagePath;
+				Package[] packs;
+				Path[] paths;
+				try {
+					auto local_package_file = list_path ~ LocalPackagesFilename;
+					logDiagnostic("Looking for local package map at %s", local_package_file.toNativeString());
+					if( !existsFile(local_package_file) ) continue over_repos;
```

Comment:
making sure to continue the loop instead of returning! (A lot of time debugging my stupidity here...)